### DISCUSSION TITLE
should a client initiating a disconnection from a server also trigger that client's disconnect event handler?

### DISCUSSION THREAD

**Discussion by arnodunstatter** [miguelgrinberg/python-socketio#1347] [Q&A]
I noticed that when my client calls its `.connect()` method it also has its connect event handler triggered, and in the [documentation for v4 ](https://python-socketio.readthedocs.io/en/v4/client.html#defining-event-handlers) (I'm using 4.6.1) it says "The connect, connect_error and disconnect events are special; they are invoked automatically when a client connects or disconnects from the server:" & "The disconnect handler is invoked for application initiated disconnects, server initiated disconnects, or accidental disconnects". That said, when my client calls its `.disconnect()` method I do not see the printout from my client's disconnection handler as I would expect. 

To try and figure out if this was arising out of the fact that the server code is written in javascript by other developers (i.e. it's a complex system) I tried to simulate a simplified version of the connection by running a pure python-socketio client-server connection simulation by using videos 1 and 10 from your [youtube python-socketio tutorial playlist](https://www.youtube.com/playlist?list=PLCuWRxjbgFnPZTBMYbz9UNGvTLNggRMjb) (great videos by the way, very helpful). There again, when my client disconnects it does not have its own disconnect event handler triggered. 

I further noticed that if I reinstall python-socketio without specifying my version such that the most up to date version (currently v5.11.2) is installed, then I see what I expect from the documentation: when my client disconnects it's disconnection event handler is also triggered.

When I look at the [source code for the JavaScript socket.js ](https://github.com/socketio/socket.io/blob/2.x/lib/socket.js) implementation for the version compatible with 4.6.1, its `Socket.prototype.onpacket` function's `case parser.DISCONNECT:` (what I believe handles incoming disconnect events) calls `this.ondisconnect()` which in turn calls `this.onclose('client namespace disconnect')` which then appears to emit both a `'disconnecting'` and `'disconnect'` event, again making me suspect that my client's disconnect event handler should be getting triggered when it calls its `.disconnect()` method.

Any ideas what I could be doing wrong?
![image](https://github.com/miguelgrinberg/python-socketio/assets/61071417/904697e0-558f-4bf1-9a51-261aaf8cf64c)

**Comment by miguelgrinberg** [comment]
You are not doing anything wrong. There is a bug in the older releases of python-socketio. This issue was fixed in release 5.1.0, you can see it referenced in the release notes for that release.

**Reply by arnodunstatter** [reply]
Thank you!
